ec/src/hashing/curve_maps/swu/mod.rs

```diff
@@ -0,0 +1,221 @@
+        let ta = xi_t2.square() + xi_t2;
+        let num_x1 = b * (ta + <P::BaseField as One>::one());
+        let div = a * if ta.is_zero() { P::XI } else { -ta };
+        let num2_x1 = num_x1.square();
+        let div2 = div.square();
+        let div3 = div2 * div;
+        let num_gx1 = (num2_x1 + a * div2) * num_x1 + b * div3;
+
+        // 5. x2 = Z * u^2 * x1
+        let num_x2 = xi_t2 * num_x1; // same div
+
+        // 6. gx2 = x2^3 + A * x2 + B  [optimized out; see below]
+        // 7. If is_square(gx1), set x = x1 and y = sqrt(gx1)
+        // 8. Else set x = x2 and y = sqrt(gx2)
+        let gx1_square;
+        let gx1;
+        let zeta_gx1;
+
+        assert!(
+            !div3.is_zero(),
+            "we have checked that neither a or xi are zero. Q.E.D."
+        );
+        let y1: P::BaseField = {
+            gx1 = num_gx1 / div3;
+            zeta_gx1 = P::ZETA * gx1;
+            if gx1.legendre().is_qr() {
+                gx1_square = true;
+                gx1.sqrt()
+                    .expect("We have checked that gx1 is a quadratic residue. Q.E.D")
+            } else {
+                gx1_square = false;
+                zeta_gx1.sqrt().expect(
+                    "zeta * gx1 is a quadratic residue because legard is multiplicative. Q.E.D",
+                )
+            }
+        };
+
+        // This magic also comes from a generalization of [WB2019, section 4.2].
+        //
+        // The Sarkar square root algorithm with input s gives us a square root of
+        // h * s for free when s is not square, where h is a fixed nonsquare.
+        // In our implementation, h = ROOT_OF_UNITY.
+        // We know that Z / h is a square since both Z and h are
+        // nonsquares. Precompute theta as a square root of Z / ROOT_OF_UNITY.
+        //
+        // We have gx2 = g(Z * u^2 * x1) = Z^3 * u^6 * gx1
+        //                               = (Z * u^3)^2 * (Z/h * h * gx1)
+        //                               = (Z * theta * u^3)^2 * (h * gx1)
+        //
+        // When gx1 is not square, y1 is a square root of h * gx1, and so Z * theta *
+        // u^3 * y1 is a square root of gx2. Note that we don't actually need to
+        // compute gx2.
+
+        let y2 = P::XI_ON_ZETA_SQRT * xi_t2 * point * y1;
+        let num_x = if gx1_square { num_x1 } else { num_x2 };
+        let y = if gx1_square { y1 } else { y2 };
+
+        let x_affine = num_x / div;
+        let y_affine = if Self::parity(&y) { -y } else { y };
+        let point_on_curve = GroupAffine::<P>::new(x_affine, y_affine, false);
+        assert!(
+            point_on_curve.is_on_curve(),
+            "swu mapped to a point off the curve"
+        );
+        Ok(point_on_curve)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ark_test_curves::bls12_381::{Fq, Fq2, Fq6};
+
+    impl ElementParity<Fq> for Fq {}
+    impl ElementParity<Fq2> for Fq2 {}
+    impl ElementParity<Fq6> for Fq6 {}
```

Comment:
Hmm these impls seem unnecessary; does `SWUMap::parity` not suffice for this?

Reply:
I am not sure how this would work - `SWUMap` is defined over `SWUParams`, and I think we would need to implement the parameters for each base field we want to test, like it is done [here](https://github.com/w3f/arkworks-algebra/blob/hash_to_curve_wb/ec/src/hashing/tests.rs#L115). That would bloat the test code in the swu module.

I agree that the above 3 lines look a little unusual, but I think it's a neat hack that leverages the fact that the `ElementParity` trait is not tied to SWUMap, and that luckily in this scope the trait is local, so we don't have to worry about orphan traits.

Correct me if I'm wrong, I'd happily integrate another solution.

Reply:
I guess I don't fully see the necessity of defining a separate trait for this, especially if we might run into orphan impl conflicts somehow.

Reply:
Done in @https://github.com/w3f/arkworks-algebra/pull/35/commits/80082c447e8d183711972b8bbc01db0c53fb3b63